File: API/src/TokenLog.cs
using System.Collections.Generic;
using API.Exceptions;

namespace API
{
    public class TokenLog
    {
        private static List<Token> tokens = new List<Token>(); 

        /// <summary>
        /// Returns true if tokens contain a string equal to tokenstring, false otherwise.
        /// </summary>
        public static bool ContainsToken(string tokenstring)
        {
            foreach (Token token in tokens)
            {
                if (tokenstring.Equals(token.GetToken()))
                    return true;
            }
            return false;
        }

        /// <summary>
        /// Returns the id of token owner, given a token key.
        /// </summary>
        public static int GetTokenOwner(string token)
        {
            foreach (Token t in tokens)
            {
                if (t.GetToken().Equals(token))
                    return t.GetOwnerId();
            }

            return -1;
        }

        /// <summary>
        /// Adds a token to list if it is unique. Returns a new token if user did not have a token, otherwise returns 
        /// the already granted token.
        /// </summary>
        public static Token AddToken(Token t)
        {
            if (ContainsToken(t.GetToken()))
                throw new TokenInUse();

            foreach(Token token in tokens)
            {
                if (t.GetOwnerId().Equals(token.GetOwnerId()))
                    return token;
            }

            tokens.Add(t);
            return t;
        }
    }
}

File: api/tokenLog.js
var tokenLog = [];
const TOKEN_LENGTH = 12;

/*
Returns a token object = {
  token: '12-char random string',
  owner: user_id
}
*/
function create_token(user_id){
  return new Promise(function(resolve){
    let token;
    do {
      token = "";
      var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
      for (var i = 0; i < TOKEN_LENGTH; i++){
        token += possible.charAt(Math.floor(Math.random() * possible.length));
      }
    } while(contains_token(token));
    resolve({token: token, owner: user_id});
  });
}

/* Returns true and deletes token from log, if owner and tokenString match.
Otherwise, tokenLog is not touched and function returns false.*/
function delete_token(tokenString, owner){
  let index = -1;
  for (var i = 0; i < tokenLog.length; i++){
    if (tokenLog[i].token === tokenString && tokenLog[i].owner === owner){
      index = i;
      break;
    }
  }
  if (index !== -1){
    tokenLog.splice(index, 1);
    return true;
  }
  return false;
}

function add_token(token){
  tokenLog.push(token);
}

function clear(){
  tokenLog = [];
}

function contains_token(tokenString){
  return tokenLog.some(function (t) {
    return t.token === tokenString;
  });
}

/*returns owner of the token, if token is present in tokenLog.*/
function get_token_owner(tokenString){
  for (var i = 0; i < tokenLog.length; i++){
    if (tokenLog[i].token === tokenString){
      return tokenLog[i].owner;
    }
  }
  return -1;
}

module.exports = {
  create_token: create_token,
  delete_token: delete_token,
  clear: clear,
  contains_token: contains_token,
  add_token: add_token,
  get_token_owner: get_token_owner
};
